remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,804 @@
+  auto val2 = I.getOperand(1);
+  if (exclude.count(val1) || exclude.count(val2)) {
+    exclude.insert(&I);
+    return VisitResult::Progress;
+  } else {
+    // FIXME(tim): is there a way to make this code more succinct and avoid duplication?
+    if (auto cint1 = llvm::dyn_cast<llvm::ConstantInt>(val1)) {
+      auto ptr = state_offset.find(val2);
+      if (ptr == state_offset.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        uint64_t offset = 0;
+        if (!GetUnsignedOffset(cint1->getSExtValue(), static_cast<int64_t>(ptr->second),
+              op, static_cast<int64_t>(state_slots.size()), &offset)) {
+          LOG(WARNING) << "Out of bounds " << (op == OpType::Plus ? "add " : "sub ")
+            << "operation: " << LLVMThingToString(&I);
+          return VisitResult::Error;
+        }
+        state_offset.emplace(&I, offset);
+        return VisitResult::Progress;
+      }
+    } else if (auto cint2 = llvm::dyn_cast<llvm::ConstantInt>(val2)) {
+      auto ptr = state_offset.find(val1);
+      if (ptr == state_offset.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        uint64_t offset = 0;
+        if (!GetUnsignedOffset(static_cast<int64_t>(ptr->second), cint2->getSExtValue(),
+              op, static_cast<int64_t>(state_slots.size()), &offset)) {
+          LOG(WARNING) << "Out of bounds " << (op == OpType::Plus ? "add " : "sub ")
+            << "operation: " << LLVMThingToString(&I);
+          return VisitResult::Error;
+        }
+        state_offset.emplace(&I, offset);
+        return VisitResult::Progress;
+      }
+    } else {
+      // check if both are in the state_offset
+      auto ptr1 = state_offset.find(val1);
+      auto ptr2 = state_offset.find(val2);
+      if (ptr1 == state_offset.end() || ptr2 == state_offset.end()) {
+        return VisitResult::NoProgress;
+      } else {
+        uint64_t offset = 0;
+        if (!GetUnsignedOffset(static_cast<int64_t>(ptr1->second),
+              static_cast<int64_t>(ptr2->second), op, static_cast<int64_t>(state_slots.size()), &offset)) {
+          LOG(WARNING) << "Out of bounds " << (op == OpType::Plus ? "add " : "sub ")
+            << "operation: " << LLVMThingToString(&I);
+          return VisitResult::Error;
+        }
+        state_offset.emplace(&I, offset);
+        return VisitResult::Progress;
+      }
+    }
+  }
+}
+
+// Visit a select instruction and update the offset map.
+VisitResult ForwardAliasVisitor::visitSelect(llvm::SelectInst &I) {
+  auto in_state_offset = false;
+  auto in_exclude_set = false;
+  uint64_t offset = 0;
+  // check both the true branch and the false branch
+  auto trueval = I.getTrueValue();
+  auto falseval = I.getFalseValue();
+  if (exclude.count(trueval)) {
+    in_exclude_set = true;
+  } else {
+    auto ptr = state_offset.find(trueval);
+    if (ptr == state_offset.end()) {
+      return VisitResult::NoProgress;
+    } else {
+      if (!in_state_offset) {
```

Comment:
This is never true. I think you should update this function to do:
```
auto true_ptr = state_offset.find(true_val);
auto false_ptr = state_offset.find(false_val);
auto in_exclude_set = exclude.count(true_val) || exclude.count(false_val);
auto in_state_offset = true_ptr != state_offset.end() || false_ptr != state_offset.end();
```

Then handle the cases accordingly.